File: python/mirage/core/wandb/pathing.py
from mirage.accessor.wandb import WandbAccessor
from mirage.core.wandb.types import RunVariables
from mirage.types import PathSpec
from mirage.utils.errors import enoent

LEAVES = ("run.json", "config.json", "summary.json", "history.jsonl")


def parts(accessor: WandbAccessor, path: PathSpec) -> list[str]:
    key = path.mount_path.strip("/")
    result = key.split("/") if key else []
    if any(not safe_name(p)
           for p in result) or (result
                                and result[0] not in accessor.config.entities):
        raise enoent(path)
    return result


def safe_name(name: str) -> bool:
    return bool(name) and name not in (".", "..") and not any(
        c in name for c in ("/", "\\", "\x00"))


def run_vars(segments: list[str]) -> RunVariables:
    entity, project, run = segments[:3]
    return {"entity": entity, "project": project, "run": run}

File: typescript/packages/core/src/core/wandb/pathing.ts
import type { WandbAccessor } from '../../accessor/wandb.ts'
import type { PathSpec } from '../../types.ts'
import { enoent } from '../../utils/errors.ts'
import type { RunVariables } from './types.ts'

export const LEAVES = ['run.json', 'config.json', 'summary.json', 'history.jsonl']
export function safeName(name: string): boolean {
  return (
    name.length > 0 &&
    !['.', '..'].includes(name) &&
    !['/', '\\', '\0'].some((char) => name.includes(char))
  )
}
export function parts(accessor: WandbAccessor, path: PathSpec): string[] {
  const key = path.mountPath.replace(/^\/+|\/+$/g, '')
  const result = key ? key.split('/') : []
  if (
    result.some((p) => !safeName(p)) ||
    (result.length > 0 && !accessor.config.entities.includes(result[0] ?? ''))
  )
    throw enoent(path)
  return result
}
export function runVars(segments: string[]): RunVariables {
  const [entity, project, run] = segments
  if (entity === undefined || project === undefined || run === undefined)
    throw new Error('W&B run path requires three components')
  return { entity, project, run }
}
